kasa/discover.py

```diff
@@ -88,29 +97,61 @@ def do_discover(self) -> None:
 
     def datagram_received(self, data, addr) -> None:
         """Handle discovery responses."""
-        ip, port = addr
-        if (
-            ip in self.discovered_devices
-            or ip in self.unsupported_devices
-            or ip in self.invalid_device_exceptions
-        ):
-            return
+        asyncio.create_task(self.datagram_received_async(data, addr))
 
+    async def datagram_received_async(self, data, addr) -> None:
+        """Handle discovery responses."""
+        ip, port = addr
+        # Prevent multiple entries due multiple broadcasts
+        async with self.ip_sem:
+            if ip in self.ip_set:
+                return
+            self.ip_set.add(ip)
+
+        protocol: Optional[TPLinkProtocol] = None
+        info: Dict
         if port == self.discovery_port:
             info = json_loads(TPLinkSmartHomeProtocol.decrypt(data))
             _LOGGER.debug("[DISCOVERY] %s << %s", ip, info)
-
         elif port == Discover.DISCOVERY_PORT_2:
             info = json_loads(data[16:])
-            self.unsupported_devices[ip] = info
-            if self.on_unsupported is not None:
-                asyncio.ensure_future(self.on_unsupported(info))
-            _LOGGER.debug("[DISCOVERY] Unsupported device found at %s << %s", ip, info)
-            if self.discovered_event is not None and "255" not in self.target[0].split(
-                "."
+            if not (
+                self.auth_credentials is not None
+                and Discover._is_klap_discovery_info(info)
             ):
-                self.discovered_event.set()
-            return
+                _LOGGER.debug("Unsupported device found at %s << %s", ip, info)
+                self.unsupported_devices[ip] = info
+                if self.on_unsupported is not None:
+                    asyncio.ensure_future(self.on_unsupported(info))
+                if self.discovered_event is not None and "255" not in self.target[
+                    0
+                ].split("."):
+                    self.discovered_event.set()
+                return
```

Comment:
Is this really necessary as it reads rather hacky?

Reply:
It's consistent with how I did it in the add new discovery protocol PR which has this check in a few places.  The thinking behind it is that the discovery done event is only triggered when targeting a host rather than a broadcast.